### ISSUE TITLE
"visit" API call looks to access the url twice

### ISSUE THREAD

**Issue by kakusuke** [cobrateam/splinter#241]
Hi,

I started a django dev server, and accessed the top page of the server using by django shell:

```
from splinter.browser import Browser
browser = Browser('phantomjs')
browser.visit('http://localhost:8080/')
```

server log is:

```
[04/Jul/2013 20:31:44] "GET / HTTP/1.1" 200 0
[04/Jul/2013 20:31:44] "GET / HTTP/1.1" 200 0
```

Why did the "visit" method call access the page twice?
It looks bit strange to me.

thanks.

**Comment by danielward** [comment]
From what I can tell, this appears to be something that is ongoing. In the meantime, I did a little digging to find that 'HTTP_ACCEPT' is only requested by the browser, not the prior request made by Splinter.

As a result, if you have a 'TEST_RUN' flag or similar for your test runner settings in Django, I found the below to work in the dispatch method of my view class prior to calling the super method (could be handled in a middleware alternatively).

So far, thing seem to be ok, though I can see how this could be an issue depending on how it is implemented. Nevertheless, this has helped me in the interim.

```python
if settings.TEST_RUN and 'HTTP_ACCEPT' not in request.META:
    return HttpResponse()
```

**Comment by ktoh** [comment]
I would suggest:

1. On the initial status code call, add a header flag to the request so that it can be distinguished server-side from the following call.  That way, server-side test harnesses can detect the status check and handle appropriately if necessary.  This should also allow for backward compatibility.

2. Allow an optional ```disable_status_check``` flag to disable the status code pre-check.  It should default to ```False``` to allow for backward compatibility.

Implementing #1 would be very easy, and not dependent on implementing #2. 
What do folks think?

**Comment by ktoh** [comment]
While this issue remains open, here is one way to distinguish the two requests by inspecting their headers:

1. The first request sets the request header to ```User-Agent = python/splinter```.  The code is [here](https://github.com/cobrateam/splinter/blob/c1c1cff513d60333e089b1a9a85d24c0099faf5a/splinter/request_handler/request_handler.py#L52).

2. The second request uses the browser's standard ```User-Agent``` (e.g. Mozilla, Chrome or whatever).

If you need to handle the first call separately server-side, you can simply inspect the request header to look for ```python/splinter``` and handle accordingly.  This isn't a solution to the issue, but may be a workaround until a solution is implemented.

**Comment by danielward** [comment]
My previous reply touched on a distinguishable difference between Splinter status checks and browser based requests. Still stands up, but it may be that I hit a snag yet.

I personally prefer the idea of an additional argument on the visit call to disable the pre-checking conditionally, as it may be that the pre-checks are useful in many cases. This would also prevent additional intervention server-side for the purposes of catering to Splinter.

**Comment by ktoh** [comment]
@danielward I agree with you completely.  My comment was poorly phrased....the optional disablement (default to False for backward compatibility) is the right solution here IMO.

**Comment by andrewsmedina** [comment]
I'm working in a partial solution for this problem. The idea is to make a lazy get for status code. So, the duplicate request will happen only to get the status code. Avoiding the problem on visit.

**Comment by danielward** [comment]
Surely that's still going to result in duplicate visits, presumably with no option to disable the feature?

Without looking over the code base, is the prior status checking fairly relied upon? I'm not suggesting that it would need removing because status checking can be useful, but subsequent assertions on page content would most likely throw assertion errors if the page didn't respond as expected in any case, whether or not the status had been checked.

**Comment by andrewsmedina** [comment]
@danielward it's need to remove `status_code` attrbitute to remove the duplicated visits. We can fix it in all drivers except selenium. =/

**Comment by blueyed** [comment]
Just for reference: pytest-splinter added a way to fetch `status_code` lazily from splinter, which might be a good enough workaround for people in need of "single requests" (ref: https://github.com/pytest-dev/pytest-splinter/commit/c51de76f93a7a8e2a22fe2e02e160045bb2bfd4f).
pytest-splinter used to not make the `status_code` available previously, to not trigger the 2nd request.

**Comment by ikeeip** [comment]
If you don't need status_code you can use the following workaround:

    from splinter.browser import Browser
    browser = Browser()
    browser.evaluate_script('document.location="http://localhost:8080/"')

**Comment by andrewsmedina** [comment]
Today I have pushed a commit that is a partial solution for this problem. Now, splinter process status_code only when it is needed. So, the duplicate request will happen only to get the status code. Avoiding the problem on visit.

**Comment by Goutte** [comment]
I lost my entire day and lots of hair on this ; I understand you can't do anything about this awful A4-sized paper cut, but maybe print/log loudly when a second request is silently done, blaming selenium's core team's lack of pragmatism ?

**Comment by andrewsmedina** [comment]
@Goutte the last splinter version, do not access the url twice on "visit". The all problem is not fixed, but is minimized.

**Comment by Goutte** [comment]
Yeah, I was using the status code the line after the `visit`, and there was a discrepancy between the page I got during the call to `visit` (which was OK) and the status code I got afterwards (which was 403).
Looks like the status code second request bypasses my authentication mechanism, somehow.

I looked in all the wrong places, of course, so I'm also to blame. :)
But with such a caveat, it's very easy to look in the wrong places.

Besides, REST nowadays makes us want to know these headers even in Gherkin-based browser-slaving feature suites, and furthermore I wanna spec 404s and 403s on regular pages, and I don't care if I can't spec 301s that way because the bowser moves to another 200 (or anything else) page right away. I can assert for success, it's good enough. If I wanna test the 301 itself, I'll make an unit test the old way.

_rambling in the wrong places_
